producer/writer/consumer_writer.go

```diff
@@ -0,0 +1,317 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package writer
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3msg/generated/proto/msgpb"
+	"github.com/m3db/m3msg/protocol/proto"
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+
+	"github.com/uber-go/tally"
+)
+
+const (
+	defaultRetryForever = true
+)
+
+var (
+	defaultConn = new(net.TCPConn)
+)
+
+type consumerWriter interface {
+	// Write writes the marshaler, it should be thread safe.
+	Write(m proto.Marshaler) error
+
+	// Init initializes the consumer writer.
+	Init()
+
+	// Close closes the consumer writer.
+	Close()
+}
+
+type consumerWriterMetrics struct {
+	ackError          tally.Counter
+	decodeError       tally.Counter
+	encodeError       tally.Counter
+	resetConn         tally.Counter
+	resetError        tally.Counter
+	connectError      tally.Counter
+	setKeepAliveError tally.Counter
+}
+
+func newConsumerWriterMetrics(scope tally.Scope) consumerWriterMetrics {
+	return consumerWriterMetrics{
+		ackError:          scope.Counter("ack-error"),
+		decodeError:       scope.Counter("decode-error"),
+		encodeError:       scope.Counter("encode-error"),
+		resetConn:         scope.Counter("reset-conn"),
+		resetError:        scope.Counter("reset-conn-error"),
+		connectError:      scope.Counter("connect-error"),
+		setKeepAliveError: scope.Counter("set-keep-alive-error"),
+	}
+}
+
+type connectFn func(addr string) (net.Conn, error)
+
+type consumerWriterImpl struct {
+	sync.RWMutex
+
+	// encodeLock controls the access to the encode function.
+	encodeLock sync.Mutex
+	// decodeLock controls the access to the decode function.
+	decodeLock sync.Mutex
+	encdec     proto.EncodeDecoder
+
+	addr            string
+	retrier         retry.Retrier
+	router          ackRouter
+	opts            Options
+	resetDelayNanos int64
+	logger          log.Logger
+
+	lastResetNanos int64
+	closed         bool
+	doneCh         chan struct{}
+	resetCh        chan struct{}
+	closeWG        sync.WaitGroup
+	m              consumerWriterMetrics
+
+	nowFn     clock.NowFn
+	connectFn connectFn
+}
+
+// TODO: Remove the nolint comment after adding usage of this function.
+// nolint: deadcode
+func newConsumerWriter(
+	addr string,
+	router ackRouter,
+	opts Options,
+) consumerWriter {
+	if opts == nil {
+		opts = NewOptions()
+	}
+	w := &consumerWriterImpl{
+		encdec:          proto.NewEncodeDecoder(defaultConn, opts.EncodeDecoderOptions()),
```

Comment:
I did it this way to reuse `resetWithConnectFn()` and `signalReset()`, it's fine if the initial constructor could not create a working connection, it will be retried in the background anyway.

Reply:
In that case can we just set connection to nil? I'm not even sure if the connection returned by `new(net.TCPConn)` is a valid object or not.

Reply:
I can't use nil directly because that would cause a encode or decode function to panic on the nil `ReadWriter`. From my test the new(net.TCPConn) just returns errors on all calls. But I could change it something like a `ErrorReadWriter` instead.